### PR TITLE
InitVar is now properly type-checked for dataclass

### LINKED ISSUE #12: Respect InitVars in type_checked_constructor
## Problem

The InitVar feature of dataclasses is not compatible with undictify.

## Possible Solution

InitVar's contained type should be checked as a special case for dataclasses. This will properly type-check the constructor.

See [here](https://docs.python.org/3.7/library/dataclasses.html#init-only-variables) for an explanation of InitVar.

## Example

```python
from dataclasses import dataclass, InitVar, field
from undictify import type_checked_constructor

@type_checked_constructor()
@dataclass
class Hello:

    x: int
    y: int
    my_init_var: InitVar[str]
    z: str = field(init=False)

    def __post_init__(self, my_init_var: str):
        self.z = f"Hello, {my_init_var}"

print(Hello(x=11, y=2, my_init_var="hello"))
```

Result:

```traceback
Traceback (most recent call last):
  File "main.py", line 18, in <module>
    print(Hello(x=11, y=2, my_init_var="hello"))
  File "~/src/sandbox/fun-lsp/.venv/lib/python3.7/site-packages/undictify/_unpack.py", line 66, in wrapper
    converters)
  File "~/src/sandbox/fun-lsp/.venv/lib/python3.7/site-packages/undictify/_unpack.py", line 185, in _unpack_dict
    converters)
  File "~/src/sandbox/fun-lsp/.venv/lib/python3.7/site-packages/undictify/_unpack.py", line 246, in _get_value
    raise TypeError(f'Key {param_name} has incorrect type: '
TypeError: Key my_init_var has incorrect type: str instead of InitVar.
```

### PR DESCRIPTION
**Pull request by pappasam**
In Python dataclasses, InitVars are passed as arguments to __init__ and
__post_init__. Unfortunately, as of Python 3.7, InitVar does not pass
any type information regarding its contained type. For example, these is
not way to determine that InitVar[str] contains a str. InitVar is
currently defined as following in the Python standard library:

  class _InitVarMeta(type):
      def __getitem__(self, params):
          return self

  class InitVar(metaclass=_InitVarMeta):
      pass

As such, it is missing the __args__ attribute found on container classes
in the typing module. InitVars are passed as arguements to __init__ but
are not set there. Their type is InitVar, so we choose to simply ignore
their type (for now) and pass the value through __init__ if it is an
InitVar.

Because of this limitation, we rely on the __post_init__ method to
determine the type of an InitVar. We provide a check for __post_init__
method arguments. Any arguments passed to this method are assumed to be
InitVars defined for the data class. Their types can be annotated in
__post_init__, and the annotations are used to perform runtime type
checks on the InitVar.

Our treatment of InitVar could result in strange edge case bugs. This is
documented in the tests (see: test_known_bug_initvar_as_argument_passes)

Resolves: https://github.com/Dobiasd/undictify/issues/12

### CODE REVIEW

**Comment by Dobiasd** [comment]
Thanks a lot for this great PR. Not just the code looks very good, I also really appreciate the precise explanation.

To make sure I understand correctly:
The changes in `type_checked_constructor` improve code quality, but are not strictly needed for the `InitVar` handling, right?

**Comment by pappasam** [comment]
tl;dr the changes to `type_checked_constructor` are necessary for `InitVar` handling.

Since `InitVar` handling requires overriding the `__post_init__` method conditionally for classes that are `dataclasses` and have a `__post_init__` method, I needed to parameterize the function `wrapper` so I could use it for multiple use cases. The old method of having `signature_ctro`, `replace_init`, and `original_ctor` would have required me to create another wrapper function and several more function-local variables (`original_ctro_post_init_`, `signature_ctro_post_init`, and `replace_post_init`, or something like that.

Ultimately, a change to `type_checked_constructor` is necessary to handle the `__post_init__` override, in addition to the `__init__` override. I did a broader refactoring of `type_checked_constructor` because I wanted to reuse patterns that were already there, but not parameterized.

**Comment by pappasam** [comment]
Note: it's currently broken for 3.6. Fixing now

**Comment by pappasam** [comment]
Fixed. Tests pass on 3.6 and 3.7. We might want to configure Tox to make running these tests easier on our local machines. I'd be happy to discuss updates to the dev environment after this PR if you'd like. I love this library and want to give it some TLC if you're open to it

**Comment by Dobiasd** [comment]
> tl;dr the changes to type_checked_constructor are necessary for InitVar handling.
> ...

Thanks for the explanation. :+1:

Could you add a unit test showing the need for the changes in `type_checked_constructor`? The current unit tests in this PR also pass when I use `type_checked_constructor` from `master`. I mean, I like your changes in that function, but I'd like to understand this interesting change completely.

---

> We might want to configure Tox to make running these tests easier on our local machines. I'd be happy to discuss updates to the dev environment after this PR if you'd like. I love this library and want to give it some TLC if you're open to it

I've not yet used tox, but I'm open to learning it. :slightly_smiling_face:
However, I already have something for testing the library locally with different python versions, based on docker compose. I could clean this up a bit and commit. Maybe it already suffices.

**Comment by Dobiasd** [comment]
Just committed the [docker-compose based tests](
https://github.com/Dobiasd/undictify/commit/66c06c70629eac84ea55b121e1ad782dd9f326b9). Just run `./run_tests_in_docker.sh` from the base directory. Then this PR is finished we can discuss if something Tox-based would be better.

**Comment by Dobiasd** [comment]
Ah, [now](https://github.com/Dobiasd/undictify/pull/13/commits/ea6d6c4394075baddb55fca0689ccb7a5cad961e) I see! Thank you so much. :heart:

**Comment by pappasam** [comment]
DO NOT MERGE. Sadly, this PR doesn't work for Python 3.8 currently. They changed the InitVar implementation so that Types can actually be checked in the InitVar itself:

```python
class _InitVarMeta(type):
    def __getitem__(self, params):
        return InitVar(params)

class InitVar(metaclass=_InitVarMeta):
    __slots__ = ('type', )

    def __init__(self, type):
        self.type = type

    def __repr__(self):
        if isinstance(self.type, type):
            type_name = self.type.__name__
        else:
            # typing objects, e.g. List[int]
            type_name = repr(self.type)
        return f'dataclasses.InitVar[{type_name}]'
```

Old implementation (for reference);

```python
class _InitVarMeta(type):
    def __getitem__(self, params):
        return self

class InitVar(metaclass=_InitVarMeta):
    pass
```

I'll need to create a check for Python 3.7 versus 3.8. Seems like the design of InitVar was an afterthought for the dataclasses team.

**Comment by pappasam** [comment]
This also tells me that the current strategy of differentiating between Python 3.7+ versus everything else may need to be reconsidered.

**Comment by pappasam** [comment]
That said, this library doesn't officially say it supports 3.8. So if you want to merge now and let us consider Python 3.8 compatibility in a future PR, that works for me.

**Comment by Dobiasd** [comment]
It's great that they "fixed" the `InitVar` problem with 3.8. But it's not so nice we now have to maintain special handling for 3.7.

Yes, I think we should support 3.8, and then have the appropriate tests.

If we would merge now, `master` would still be 3.8 compatible, just not for `InitVar`, right? If so, I think it's totally fine to merge and add `InitVar` 3.8 support and tests in a new PR.

**Comment by pappasam** [comment]
@Dobiasd I've added InitVar parsing support for Python 3.8. Tests now pass for 3.6, 3.7, and 3.8. Please check the latest commit and see if you agree with it. If yes, for the next release, I'd also bump this library's official support for 3.8 in your setup.py.

**Comment by pappasam** [comment]
Please do not merge yet. Found another weird edge case involving lists, will clarify when I have time

**Comment by Dobiasd** [comment]
Oh, I was just about to send the following. :grin:

> That looks very good to me. :+1:
> Yes, I'd guess we could increase the version number in `setup.py` from `0.6.5` to `0.7.0` since we > added a new feature and not just a patch. Would you agree?
> If settled, we can merge.
> Looking forward to the Python 3.8 tests in a separate PR. :slightly_smiling_face:

OK, just let me know when you feel confident that every relevant thing is covered.

**Comment by pappasam** [comment]
The problem is fixed AFAIK. Check out the latest commit / message. I'd play around with it yourself for a little while to see if you notice anything that I may have overlooked.

Regarding setup.py, I agree with your suggestion and would add the following:

```python
setuptools.setup(
    name="undictify",
    version="0.7.0",
    author="Tobias Hermann",
    author_email="editgym@gmail.com",
    description="Type-checked function calls at runtime",
    long_description=long_description,
    long_description_content_type="text/markdown",
    url="http://github.com/Dobiasd/undictify",
    packages=setuptools.find_packages(),
    classifiers=(
        "Programming Language :: Python :: 3.6",
        "Programming Language :: Python :: 3.7",
        "Programming Language :: Python :: 3.8",
        "License :: OSI Approved :: MIT License",
        "Operating System :: OS Independent",
    ),
)
```

3.8 support, and minor version bump.

**Comment by Dobiasd** [comment]
Looks good with the tests. If you do the changes in `setup.py` too (and remove the debug `print` in the code) we should be ready to merge and release. In case we find a bug later, we can still commit a patch. :)

If you like, you can also already enable Python 3.8 in the travis config.

**Comment by pappasam** [comment]
Done. I'm not super familiar with Travis, so please verify that the Travis commit is to your liking.

**Comment by Dobiasd** [comment]
Looks good. :+1:

I wonder why the Travis integration is not shown here in the web interface of this pull request. I did not change the webhook settings, and it [was working for an earlier PR](https://i.imgur.com/ffJEURm.png). `¯\_(ツ)_/¯`

Anyways, because they are not shown, I only looked into them now. You can find the results of the latest run here: https://travis-ci.org/Dobiasd/undictify/builds/606912956

It's failing only because of some complaints `pylint` and `mypy` are having.
To make the feedback loop when trying to fix them quicker, I recommend running the tests locally (`./run_tests_in_docker.sh` or `./test/run_code_checks.sh || ./test/run_code_checks_3_7_and_up.sh`).

Once fixed, I guess we are ready to squash / force push and merge. :slightly_smiling_face:

**Comment by pappasam** [comment]
Regarding mypy / pylint / docker-compose:

1. Mypy and pylint aren't pinned. These libraries have been changing at a rapid pace over the past couple months, so their output will differ over time unless we pick specific versions for testing purposes.
2. When I run these tools locally, I receive different error messages than you do. We need to commit a .pylintrc and a mypy.ini if the output of these tools is important for PR's.
3. "./run_tests_in_docker.sh" does not work because there is no docker-compose.yml committed to this repository.

These issues should be handled by you. I have my preferences, but if you let me run with them, I'll probably set up Poetry, run all the code through black, and set up pre-commit hooks :p

**Comment by Dobiasd** [comment]
Oh, sorry about the missing `docker-compose.yml`. I simply forgot to remove it from the `.gitignore`. [Now](https://github.com/Dobiasd/undictify/commit/23834b1761ffad0e98ea9da7b836046999b38659) it should be working.

Regarding the possibly new errors with new versions of `pylint` and `mypy`: I'm aware of that, but up to now it happened only very rarely, so it has not yet been a big issue for me.

My choice for docker compose wasn't really well-researched. I just quickly went with what I already knew. I'm totally happy about learning something new/better. If we want to discuss this in a separate issue or PR, would you think we are ready to merge this one? If you prefer, I could handle the linter messages on `master` afterward.

**Comment by pappasam** [comment]
IMO, this PR is ready to be merged. Let's handle outstanding styling / linting issues in a new PR.

If you'd like, we can make the combination of InitVar handling + styling fixes the 0.7.0 release.

**Comment by Dobiasd** [comment]
OK, cool. Thanks so far already for this very good improvement of the library. :heart:

I'll commit the linter fixes soon.

**Comment by Dobiasd** [comment]
I've calmed down the linters and released version `0.7.0`. :tada:
